rdm/audit_for_gaps.py

```diff
@@ -0,0 +1,133 @@
+    raw_checklists = list(_read_raw_checklists(checklist_sources))
+    include_files, reduced_checklist = _split_out_include_files(raw_checklists)
+    if include_files:
+        unread_files = include_files.difference(already_included)
+        already_included = already_included.union(include_files)
+        return reduced_checklist + list(_read_checklists(_checklist_generator(unread_files), already_included))
+    else:
+        return reduced_checklist
+
+
+def _read_raw_checklists(checklist_sources):
+    for checklist_text, path in checklist_sources:
+        yield from _flat_file_parser(checklist_text, path)
+
+
+def _flat_file_parser(checklist_text, path):
+    for line_text in checklist_text.split('\n'):
+        yield from _parsed_line(line_text, path)
+
+
+def _parsed_line(line_text, path):
+    if line_text:
+        tokens = line_text.split(' ')
+        if tokens:
+            key = tokens[0]
+            remainder = ' '.join(tokens[1:])
+            if not key.startswith('#'):
+                if key == 'include':
+                    yield {'include': remainder, 'path': path}
+                else:
+                    yield {'reference': key, 'description': remainder}
+
+
+def _split_out_include_files(checklist):
+    include_files = set()
+    reduced_checklist = []
+    for item in checklist:
+        include_file = item.get('include')
+        if include_file:
+            path = item.get('path')
+            if path:
+                include_file = path + '/' + include_file
+            include_files.add(include_file)
+        else:
+            reduced_checklist.append(item)
+    return include_files, reduced_checklist
+
+
+def _extract_keys_from_checklist(checklist):
+    for item in checklist:
+        key = item.get('reference')
+        if key:
+            yield key
+
+
+def _find_keys_in_sources(source_generator, checklist_keys):
+    for content in source_generator:
+        yield from _find_keys_in_content(content, checklist_keys)
+
+
+def _find_keys_in_content(content, checklist_keys):
+    for key in checklist_keys:
+        if key in content:
+            yield key
+
+
+def _report_failures(failing_checklists):
+    # Note output conforms to checklist format, so can be used as a checklist itself.
+    failure_count = len(failing_checklists)
+    plural = 's' if failure_count > 1 else ''
+    print(f'# Failed {failure_count} item{plural}:')
+    unsorted_failures = []
+    for failing_checklist_item in failing_checklists:
+        key = failing_checklist_item.get('reference', '')
+        description = failing_checklist_item.get('description', '')
+        unsorted_failures.append(f'{key} {description}')
+    sorted_failures = sorted(unsorted_failures)
+    for line in sorted_failures:
+        print(line)
+
+
+def _report_success():
+    # Note output conforms to checklist format, so can be used as a checklist itself.
+    print("# Success: all checklists passed.")
```

Comment:
Perhaps it would be helpful to print out all of the checks that passed, and which lines in which files they passed from?

Reply:
If this is more than a couple hours of work, maybe we should skip for now.

Reply:
That would be pretty noisy. As is the failing list gradually shrinks as the user makes improvements until it is down to the single success line. However, a command line option to parse and sort the checklists, without documents, then printing the full list could be very useful for debugging checklists.

Reply:
Lets hold off on making any changes to the output format for now... I think we will eventually want to produce a checklist we can give to auditors, which would show every item completed and where in the documents each requirement is fulfilled. This could get complicated, since really we will want the document and section number.

I don't know how often this will come up in practice, so we should wait until we know more about what format auditors like/expect.

Reply:
Because it was simple I added the option to show just the checklist content. 

I agree that a kind of indexed output would be a useful extension. It could start with the checklist derived keys and then generate an index for the fully formatted document. It would need to skip items not in a particular document and allow for multiple references as the auditor may need to see all of them to understand the authors intent. The extension probably needs to be outside of the gap command but could borrow the internal methods like extracting the keys.

Reply:
> Because it was simple I added the option to show just the checklist content.

Great! I left a comment below about a possible simplification related to this.

>  It would need to skip items not in a particular document and allow for multiple references as the auditor may need to see all of them to understand the authors intent. The extension probably needs to be outside of the gap command but could borrow the internal methods like extracting the keys.

Absolutely!